### PR TITLE
Engine API Client Authentication for the Merge via HTTP

### PR DESCRIPTION
**Pull request by rauljordan**
This PR adds support for JWT authentication for the Engine API used for the Ethereum "Merge". The specification for how auth should work is mentioned [here](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md). This PR includes the following changes:
- [x] Addition of JWT authorization tokens in HTTP headers when making HTTP requests to the engine API endpoint
- [x] Ensuring we use HMAC HS256 for the token string generation
- [x] Giving Prysm beacon nodes the `--jwt-secret` flag, which defines a path to a file containing a hex-encoded, 32 byte secret to be used when generating the JWT token string. See the requirements [here](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution)
- [x] Tests to ensure we are using a valid JWT with claims as the spec needs, via round-trip HTTP tests
- [x] Tests to ensure the parsing and validation of the jwt-secret file provided by the user via a flag is aligned with spec requirements

**NOTE**: This PR also updates our JWT library to v4.3.0, which is the latest [release](https://github.com/golang-jwt/jwt/releases/tag/v4.3.0)

### CODE REVIEW

File: beacon-chain/powchain/engine-api-client/v1/client.go [hunk 1]
```diff
@@ -65,6 +65,11 @@ func New(ctx context.Context, endpoint string, opts ...Option) (*Client, error)
 	c := &Client{
 		cfg: defaultConfig(),
 	}
+	for _, opt := range opts {
```
Comment (rauljordan) [inline beacon-chain/powchain/engine-api-client/v1/client.go:68]: We want to configure options before we dial, not after

File: cmd/beacon-chain/powchain/options.go [hunk 2]
```diff
@@ -13,16 +19,45 @@ var log = logrus.WithField("prefix", "cmd-powchain")
...
 	if executionEndpoint != "" {
 		opts = append(opts, powchain.WithExecutionEndpoint(executionEndpoint))
 	}
+	if len(jwtSecret) > 0 {
+		opts = append(opts, powchain.WithExecutionClientJWTSecret(jwtSecret))
+	}
 	return opts, nil
 }
 
+func parseJWTSecret(c *cli.Context) ([]byte, error) {
```
Comment (rauljordan) [inline cmd/beacon-chain/powchain/options.go:39]: Cross-check vs the spec here: https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution

File: beacon-chain/powchain/engine-api-client/v1/auth.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// Implements the http.RoundTripper interface to add JWT authentication
+// support to an HTTP client used for interacting with an execution node.
+// See the specification for more details on the supported JWT claims:
+// https://github.com/ethereum/execution-apis.
+type jwtTransport struct {
+	underlyingTransport http.RoundTripper
+	jwtSecret           []byte
+}
+
+// RoundTrip --
```
Comment (prestonvanloon) [inline beacon-chain/powchain/engine-api-client/v1/auth.go:21]: Add documentation here. What is this used for?

File: beacon-chain/powchain/engine-api-client/v1/auth.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/pkg/errors"
+)
+
+// Implements the http.RoundTripper interface to add JWT authentication
+// support to an HTTP client used for interacting with an execution node.
+// See the specification for more details on the supported JWT claims:
+// https://github.com/ethereum/execution-apis.
```
Comment (prestonvanloon) [inline beacon-chain/powchain/engine-api-client/v1/auth.go:15]: Do you have a better link so readers don't have to dig around that repo? I couldn't immediately find it

File: beacon-chain/powchain/engine-api-client/v1/auth.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// RoundTrip --
+func (t *jwtTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"iat": time.Now().Unix(), // Issued at.
+	})
+	tokenString, err := token.SignedString(t.jwtSecret)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not produce signed JWT token")
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %v", tokenString))
```
Comment (prestonvanloon) [inline beacon-chain/powchain/engine-api-client/v1/auth.go:30]: Do you need sprintf for this or can you just add the strings?

File: cmd/beacon-chain/flags/base.go [hunk 6]
```diff
@@ -23,6 +23,13 @@ var (
 		Usage: "An http endpoint for an Ethereum execution node",
 		Value: "",
 	}
+	// ExecutionJWTSecretFlag provides a path to a file containing a hex-encoded, 256 bit JWT secret
+	// which will be used to authenticate with Ethereum execution nodes via an HTTP connection.
+	ExecutionJWTSecretFlag = &cli.StringFlag{
+		Name:  "jwt-secret",
+		Usage: "File path to a 256 bit, hex-encoded JWT secret used for authenticating with Ethereum execution nodes via HTTP",
```
Comment (prestonvanloon) [inline cmd/beacon-chain/flags/base.go:30]: Can you add more to the usage? Perhaps elaborating on some of the default behavior or error causes would be helpful for the user.

https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#key-distribution

> If such a parameter is not given, the client SHOULD generate such a token, valid for the duration of the execution, and store it the hex-encoded secret as a jwt.hex file on the filesystem. This file can then be used to provision the counterpart client.
>
> If such a parameter is given, but the file cannot be read, or does not contain a hex-encoded key of at least 256 bits, the client should treat this as an error: either abort the startup, or show error and continue without exposing the authenticated port.

File: cmd/beacon-chain/powchain/options.go [hunk 7]
```diff
@@ -13,16 +19,45 @@ var log = logrus.WithField("prefix", "cmd-powchain")
...
 	opts := []powchain.Option{
 		powchain.WithHttpEndpoints(endpoints),
 		powchain.WithEth1HeaderRequestLimit(c.Uint64(flags.Eth1HeaderReqLimit.Name)),
 	}
 	if executionEndpoint != "" {
 		opts = append(opts, powchain.WithExecutionEndpoint(executionEndpoint))
 	}
+	if len(jwtSecret) > 0 {
+		opts = append(opts, powchain.WithExecutionClientJWTSecret(jwtSecret))
+	}
```
Comment (prestonvanloon) [inline cmd/beacon-chain/powchain/options.go:35]: Shouldn't we create a secret and save it?

**Comment by prestonvanloon** [comment]
I think this PR is lacking some of the requirements to meet the spec.

Keys:
- If such a parameter is not given, the client SHOULD generate such a token, valid for the duration of the execution, and store it the hex-encoded secret as a jwt.hex file on the filesystem. This file can then be used to provision the counterpart client.
- If such a parameter is given, but the file cannot be read, or does not contain a hex-encoded key of at least 256 bits, the client should treat this as an error: either abort the startup, or show error and continue without exposing the authenticated port.

Claims:
- Required: iat (issued-at) claim. The EL SHOULD only accept iat timestamps which are within +-5 seconds from the current time.
- Optional: id claim. The CL MAY use this to communicate a unique identifier for the individual CL node.
- Optional: clv claim. The CL MAY use this to communicate the CL node type/version.
Are there plans to support the option information?

Given that tokens are only valid +/- 5 seconds from iat, should you set "exp" to the iat + 5s? https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4

**Comment by rauljordan** [comment]
This is ready again @prestonvanloon. I will open an issue in the execution-apis repo with some of our feedback to the spec as soon as I get the chance this weekend

File: cmd/beacon-chain/powchain/options.go [hunk 8]
```diff
@@ -13,16 +19,53 @@ var log = logrus.WithField("prefix", "cmd-powchain")
...
+		return nil, nil
+	}
+	enc, err := file.ReadFileAsBytes(jwtSecretFile)
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) == 0 {
+		return nil, fmt.Errorf("provided JWT secret in file %s cannot be empty", jwtSecretFile)
+	}
+	secret, err := hex.DecodeString(strings.TrimPrefix(string(enc), "0x"))
```
Comment (rauljordan) [inline cmd/beacon-chain/powchain/options.go:59]: geth's hexutil library actually errors if a string is not 0x prefixed. We want to allow some flexibility here and not be so harsh on users

File: cmd/beacon-chain/main.go [hunk 9]
```diff
@@ -231,11 +232,11 @@ func startNode(ctx *cli.Context) error {
 
 	blockchainFlagOpts, err := blockchaincmd.FlagOptions(ctx)
 	if err != nil {
-		return nil
+		return err
```
Comment (rauljordan) [inline cmd/beacon-chain/main.go:235]: huge whoops here...glad we caught it

File: cmd/beacon-chain/flags/base.go [hunk 10]
```diff
@@ -23,6 +23,14 @@ var (
 		Usage: "An http endpoint for an Ethereum execution node",
 		Value: "",
 	}
+	// ExecutionJWTSecretFlag provides a path to a file containing a hex-encoded string representing a 32 byte secret
+	// used to authenticate with an execution node via HTTP. This is required if using an HTTP connection, otherwise all requests
+	// to execution nodes for consensus-related calls will fail. This is not required if using an IPC connection.
+	ExecutionJWTSecretFlag = &cli.StringFlag{
+		Name:  "jwt-secret",
+		Usage: "REQUIRED if connecting to an execution node via HTTP. Provides a path to a file containing a hex-encoded string representing a 32 byte secret used for authentication with an execution node via HTTP. If this is not set, all requests to execution nodes via HTTP for consensus-related calls will fail, which will prevent your validators from performing their duties. This is not required if using an IPC connection.",
```
Comment (prestonvanloon) [inline cmd/beacon-chain/flags/base.go:31]: String too long, please break so it doesn't exceed 100 characters per line

**Review by prestonvanloon — COMMENTED** [review]
LGTM, just one comment

**Comment by rauljordan** [comment]
Ready again @prestonvanloon

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**